### PR TITLE
8342962: [s390x] TestOSRLotsOfLocals.java crashes

### PR DESCRIPTION
**Pull request by offamitkumar**
We are on thin ice with the `TestOSRLotsOfLocals` test. Before it breaks, I'd like to provide the fix. :-)

Testing : Tier1 test with fastdebug vm.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8342962](https://bugs.openjdk.org/browse/JDK-8342962): [s390x] TestOSRLotsOfLocals.java crashes (**Bug** - P3)

### Reviewers
 * [Lutz Schmidt](https://openjdk.org/census#lucy) (@RealLucy - **Reviewer**)
 * [Martin Doerr](https://openjdk.org/census#mdoerr) (@TheRealMDoerr - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21703/head:pull/21703` \
`$ git checkout pull/21703`

Update a local copy of the PR: \
`$ git checkout pull/21703` \
`$ git pull https://git.openjdk.org/jdk.git pull/21703/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21703`

View PR using the GUI difftool: \
`$ git pr show -t 21703`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21703.diff">https://git.openjdk.org/jdk/pull/21703.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21703#issuecomment-2437077327)

### CODE REVIEW

File: src/hotspot/cpu/s390/c1_LIRAssembler_s390.cpp [hunk 1]
```diff
@@ -131,9 +131,21 @@ void LIR_Assembler::osr_entry() {
...
+    assert(frame::interpreter_frame_monitor_size() == BasicObjectLock::size(), "adjust code below");
+
+    const int locals_space = BytesPerWord * method() -> max_locals();
+    int monitor_offset = locals_space + (2 * BytesPerWord) * (number_of_locks - 1);
+    bool handled_manually = false;
+
+    if (!Immediate::is_simm20(monitor_offset + BytesPerWord) && number_of_locks > 0) {
+      // z_lg can only handle displacement upto 20bit signed binary integer
+      __ load_const(Z_R0_scratch, locals_space);
+      __ z_algr(OSR_buf, Z_R0_scratch);
```
Comment (RealLucy) [inline src/hotspot/cpu/s390/c1_LIRAssembler_s390.cpp:144]: How about using 
`__ z_algfi(OSR_buf, locals_space);`
instead?

File: src/hotspot/cpu/s390/c1_LIRAssembler_s390.cpp [hunk 2]
```diff
@@ -147,6 +159,12 @@ void LIR_Assembler::osr_entry() {
       __ z_lg(Z_R1_scratch, slot_offset + 1*BytesPerWord, OSR_buf);
       __ z_stg(Z_R1_scratch, frame_map()->address_for_monitor_object(i));
     }
+
+    if (handled_manually) {
+      // Z_R0 is killed by asm_assert_mem8_isnot_zero
+      __ load_const(Z_R0_scratch, locals_space);
+      __ z_slgr(OSR_buf, Z_R0_scratch);
```
Comment (RealLucy) [inline src/hotspot/cpu/s390/c1_LIRAssembler_s390.cpp:166]: How about using 
`__ z_slgfi(OSR_buf, locals_space);`
instead?

**Review by RealLucy — CHANGES_REQUESTED** [review]
Just some academic improvements...

File: src/hotspot/cpu/s390/c1_LIRAssembler_s390.cpp (hunk 2 above)
Comment (offamitkumar) [reply src/hotspot/cpu/s390/c1_LIRAssembler_s390.cpp]: updated :-)

File: src/hotspot/cpu/s390/c1_LIRAssembler_s390.cpp [hunk 3]
```diff
@@ -131,9 +131,20 @@ void LIR_Assembler::osr_entry() {
...
   Register OSR_buf = osrBufferPointer()->as_register();
-  { assert(frame::interpreter_frame_monitor_size() == BasicObjectLock::size(), "adjust code below");
-    int monitor_offset = BytesPerWord * method()->max_locals() +
-      (2 * BytesPerWord) * (number_of_locks - 1);
+  {
+    assert(frame::interpreter_frame_monitor_size() == BasicObjectLock::size(), "adjust code below");
+
+    const int locals_space = BytesPerWord * method() -> max_locals();
+    int monitor_offset = locals_space + (2 * BytesPerWord) * (number_of_locks - 1);
+    bool handled_manually = false;
```
Comment (RealLucy) [inline src/hotspot/cpu/s390/c1_LIRAssembler_s390.cpp:139]: Sorry, another idea (no need to accept it):
```
  bool large_offset = ! Immediate::is_simm20(monitor_offset + BytesPerWord) && number_of_locks > 0;

  if (large_offset) {
. . .
```

**Review by RealLucy — APPROVED** [review]
Looks good now.
Yet another idea, though.

**Review by RealLucy — APPROVED** [review]
Wonderful!

**Comment by offamitkumar** [comment]
@TheRealMDoerr can I get a review for this fix :-)

**Review by TheRealMDoerr — APPROVED** [review]
Looks correct to me. I suggest testing it by setting `large_offset` to true.

**Comment by offamitkumar** [comment]
> I suggest testing it by setting large_offset to true.

I ran tier1 test again with suggested settings and didn't see any regression.

**Comment by offamitkumar** [comment]
Thanks @RealLucy, @TheRealMDoerr for the reviews. 

/integrate

**Pull request closed without merging**